packages/amplify-provider-awscloudformation/src/cloud-formation-error-handler.ts

```diff
@@ -0,0 +1,46 @@
+import { $TSAny, AmplifyError } from 'amplify-cli-core';
+import { deserializeErrorMessages, CFNErrorMessage, CFNErrorMessages } from './aws-utils/cloudformation-error-serializer';
+
+const s3Indicator = '(AWS::S3::Bucket)';
+
+const handleS3Error = (err: Error & { details?: string }): void => {
+  const deserializedErrorMessages: CFNErrorMessages = deserializeErrorMessages(err.details);
+  const bucketNameRegex = /(S3Bucket|CustomMessageConfirmationBucket) \(AWS::S3::Bucket\)*/;
+  const bucketReasonRegex = /.* already exists*/;
+  const bucketExistsLines: Array<CFNErrorMessage> = deserializedErrorMessages.messages.filter(
+    (message) => bucketNameRegex.test(message.name) && bucketReasonRegex.test(message.reason),
+  );
+  if (bucketExistsLines.length) {
+    const messageWithError: CFNErrorMessage = bucketExistsLines[0];
+    const bucketRegex = /(.*) already exists*/;
+    const bucketName = messageWithError.reason.match(bucketRegex)[1];
+    throw new AmplifyError('ResourceAlreadyExistsError', {
+      message: `The S3 bucket ${bucketName} already exists.`,
+      resolution: `Please delete this bucket in the AWS S3 console and try again. The bucket can be found at: https://s3.console.aws.amazon.com/s3/buckets/${bucketName}.`,
+    });
+  }
+};
+
+export const handleCloudFormationError = (err: Error & { details?: string }): void => {
+  if (err?.name === 'ValidationError' && err?.message === 'No updates are to be performed.') {
+    return;
+  }
+
+  if (err?.name === 'ValidationError' && (err?.message ?? '').includes('_IN_PROGRESS state and can not be updated.')) {
+    throw new AmplifyError(
+      'DeploymentInProgressError',
+      {
+        message: 'Deployment is already in progress.',
+        resolution: 'Wait for the other deployment to finish and try again.',
+        code: (err as $TSAny).code,
+      },
+      err,
+    );
+  }
+
+  if (err?.details?.includes(s3Indicator)) {
+    handleS3Error(err);
```

Comment:
much nicer and cleaner! Eventually when we run into more CFN errors that are not correctly handled we can move this to be a switch statement. Overall, awesome work!